### PR TITLE
Refactor helm init command for local helm home directory initialization reuse

### LINKED ISSUE #139: the 'home' parameter on the helm provider is broken
### Terraform Version
Terraform v0.11.10
+ provider.aws v1.41.0
+ provider.external v1.0.0
+ provider.helm v0.1.0
+ provider.kubernetes v1.3.0
+ provider.local v1.1.0
+ provider.null v1.0.0
+ provider.template v1.0.0

### Affected Resource(s)
- helm provider
- helm repository

### Terraform Configuration Files
works:

```hcl
provider "helm" {
  service_account = "tiller"
  tiller_image    = "gcr.io/kubernetes-helm/tiller:v2.11.0"
  namespace       = "kube-system"
  debug           = "true"

  kubernetes {
    config_path = "./kubeconfig_${var.release_env[var.environment]}"
  }
}

resource "helm_repository" "artifactory" {
  name = "artifactory"
  url  = "https://artifactory.granicuslabs.com/artifactory/helm"
}
```
doesn't work:
```hcl
provider "helm" {
  service_account = "tiller"
  tiller_image    = "gcr.io/kubernetes-helm/tiller:v2.11.0"
  namespace       = "kube-system"
  debug           = "true"
  home            = "./.helm"

  kubernetes {
    config_path = "./kubeconfig_${var.release_env[var.environment]}"
  }
}

resource "helm_repository" "artifactory" {
  name = "artifactory"
  url  = "https://artifactory.granicuslabs.com/artifactory/helm"
}
```

### Expected Behavior
it should initialize helm in the directory specified in 'home' and install tiller.

### Actual Behavior
```
helm_repository.artifactory: Creating...
  metadata.#: "" => "<computed>"
  name:       "" => "artifactory"
  url:        "" => "https://artifactory.granicuslabs.com/artifactory/helm"

Error: Error applying plan:

1 error(s) occurred:

* helm_repository.artifactory: 1 error(s) occurred:

* helm_repository.artifactory: Couldn't load repositories file (.helm/repository/repositories.yaml).
You might need to run `helm init` (or `helm init --client-only` if tiller is already installed)
```

so I run 
```
export HELM_HOME="./.helm"
helm init --client-only    
Creating ./.helm
Creating .helm/repository
Creating .helm/repository/cache
Creating .helm/repository/local
Creating .helm/plugins
Creating .helm/starters
Creating .helm/cache/archive
Creating .helm/repository/repositories.yaml
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com
Adding local repo with URL: http://127.0.0.1:8879/charts
$HELM_HOME has been configured at ./.helm.
Not installing Tiller due to 'client-only' flag having been set
Happy Helming!
```

THEN I run `terraform apply` again and get:
```helm_repository.artifactory: Creating...
  metadata.#: "" => "<computed>"
  name:       "" => "artifactory"
  url:        "" => "https://artifactory.granicuslabs.com/artifactory/helm"

Error: Error applying plan:

1 error(s) occurred:

* helm_repository.artifactory: 1 error(s) occurred:

* helm_repository.artifactory: Looks like "https://artifactory.granicuslabs.com/artifactory/helm" is not a valid chart repository or cannot be reached: open .helm/repository/cache/.helm/repository/cache/artifactory-index.yaml: no such file or directory```
```
in addition, helm tiller is not installed in my cluster.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. add a `home = ./.helm` to the helm provider 
2. `terraform apply`

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-1234

### PR DESCRIPTION
**Pull request by pdecat**
**What this PR does / why we need it**:

This PR refactors the helm init command to allow other programs to initialize the local helm home directory without shelling out to the `helm init` command.

The only moved parts are:
- `ensureDirectories()`
- `ensureDefaultRepos()`
- `ensureRepoFileFormat()`

My current use case is with the terraform-provider-helm to resolve https://github.com/terraform-providers/terraform-provider-helm/issues/139

**Special notes for your reviewer**:

~I've prefixed the PR title as WIP because the terraform provider part is not yet implemented (in progress) but I wanted to check early if this sort of change would be acceptable on the helm end.~

**If applicable**:
- [x] this PR contains documentation (N/A: no CLI user visible change)
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility (N/A: no structure change)

**Unit test results**:

```
# go test -v ./cmd/helm ./cmd/helm/installer -run '^Test(InitCmd|EnsureHome|Initialize)' -count=1
=== RUN   TestInitCmd
--- PASS: TestInitCmd (50.96s)
=== RUN   TestInitCmd_exists
--- PASS: TestInitCmd_exists (87.33s)
=== RUN   TestInitCmd_clientOnly
--- PASS: TestInitCmd_clientOnly (4.09s)
=== RUN   TestInitCmd_dryRun
--- PASS: TestInitCmd_dryRun (0.00s)
=== RUN   TestInitCmd_tlsOptions
--- PASS: TestInitCmd_tlsOptions (0.00s)
=== RUN   TestInitCmd_output
--- PASS: TestInitCmd_output (0.00s)
PASS
ok      k8s.io/helm/cmd/helm    142.437s
=== RUN   TestInitialize
--- PASS: TestInitialize (52.90s)
=== RUN   TestEnsureHome
--- PASS: TestEnsureHome (36.85s)
PASS
ok      k8s.io/helm/cmd/helm/installer  89.772s
```

### CODE REVIEW

**Comment by pdecat** [comment]
Note: backported on v2.11.0 for easier testing on terraform-provider-helm: https://github.com/pdecat/helm/tree/v2.11.0-backport-5189

File: cmd/helm/installer/init.go
Comment (pdecat) [inline cmd/helm/installer/init.go]: Maybe this should be named `EnsureHome()`?

**Comment by pdecat** [comment]
Remove the WIP prefix as it's working fine: https://github.com/terraform-providers/terraform-provider-helm/pull/185

File: cmd/helm/installer/init.go
Comment (bacongobbler) [reply cmd/helm/installer/init.go]: Yeah, EnsureHome would be a bit more clearer on intent.

**Comment by bacongobbler** [comment]
Nice refactor! Code here looks good to me. Would you mind rebasing this?

**Comment by pdecat** [comment]
@bacongobbler thanks!

Sure, rebased.

**Review by bacongobbler — APPROVED** [review]
Needs another maintainer to sign off since this is a `size/L`, but this looks good to merge.

In the meantime, if you want to change Initialize to EnsureHome then that'd be great, but it's not blocking this from going through.

**Comment by pdecat** [comment]
Hi @bacongobbler, 10 days have passed since you approved this PR, how can this be moved further? Am I supposed to ping another maintainer? Should I rebase once again?

**Comment by pdecat** [comment]
Rebased on master.

**Comment by bacongobbler** [comment]
Yes, please try reaching out to another maintainer to see if they have the time to review it.

**Comment by pdecat** [comment]
Hi @technosophos @mattfarina, as the most active maintainers here apart from @bacongobbler who already reviewed and approved this PR, could you please take a look at it?

**Review by technosophos — APPROVED** [review]
Reviewed code, and checked specifically for compatibility breakers. All looks great. @bacongobbler feel free to merge at your convenience.

**Comment by pdecat** [comment]
Thank you so much @bacongobbler & @technosophos !

**Pull request merged**
